src/main/java/dev/sigstore/fulcio/client/FulcioClient.java

```diff
@@ -86,43 +89,57 @@ public FulcioClient build() {
-      throws IOException, CertificateException {
-    URI fulcioEndpoint = serverUrl.resolve(SIGNING_CERT_PATH);
-
-    HttpRequest req =
-        HttpClients.newHttpTransport(httpParams)
-            .createRequestFactory()
-            .buildPostRequest(
-                new GenericUrl(fulcioEndpoint),
-                ByteArrayContent.fromString("application/json", cr.toJsonPayload()));
-
-    req.getHeaders().setAccept("application/pem-certificate-chain");
-    req.getHeaders().setAuthorization("Bearer " + cr.getIdToken());
-
-    HttpResponse resp = req.execute();
-    if (resp.getStatusCode() != 201) {
-      throw new IOException(
-          String.format(
-              "bad response from fulcio @ '%s' : %s", fulcioEndpoint, resp.parseAsString()));
-    }
-
-    String sctHeader = resp.getHeaders().getFirstHeaderStringValue("SCT");
-    try (InputStream content = resp.getContent()) {
-      var signingCert =
-          SigningCertificate.newSigningCertificate(
-              CharStreams.toString(new InputStreamReader(content, resp.getContentCharset())),
-              sctHeader);
-      if (signingCert.getDetachedSct().isEmpty() && !signingCert.hasEmbeddedSct() && requireSct) {
-        throw new IOException(
-            "no signed certificate timestamps were found in response from Fulcio");
+      throws InterruptedException, CertificateException, IOException {
+    // TODO: we should reuse this message channel across various executions of this client and
+    // add a client shutdown method instead of encapsulating it all in here
+    var channel = GrpcChannels.newManagedChannel(serverUrl, httpParams);
+
+    try {
+      var client = CAGrpc.newBlockingStub(channel);
+      var credentials = Credentials.newBuilder().setOidcIdentityToken(cr.getIdToken()).build();
+
+      String pemEncodedPublicKey =
+          "-----BEGIN PUBLIC KEY-----\n"
+              + Base64.getEncoder().encodeToString(cr.getPublicKey().getEncoded())
+              + "\n-----END PUBLIC KEY-----";
+      var publicKeyRequest =
+          PublicKeyRequest.newBuilder()
+              .setPublicKey(
+                  PublicKey.newBuilder()
+                      .setAlgorithm(cr.getPublicKeyAlgorithm())
+                      .setContent(pemEncodedPublicKey)
+                      .build())
+              .setProofOfPossession(ByteString.copyFrom(cr.getProofOfPossession()))
+              .build();
+      var req =
+          CreateSigningCertificateRequest.newBuilder()
+              .setCredentials(credentials)
+              .setPublicKeyRequest(publicKeyRequest)
+              .build();
+
+      var certs = client.createSigningCertificate(req);
+
+      if (certs.getCertificateCase()
+          == dev.sigstore.fulcio.v2.SigningCertificate.CertificateCase
+              .SIGNED_CERTIFICATE_DETACHED_SCT) {
+        if (certs.getSignedCertificateDetachedSct().getSignedCertificateTimestamp().isEmpty()
+            && requireSct) {
+          throw new IOException(
```

Comment:
IOException is for unexpected network/file system errors isn't it?  This looks like serialization worked fine but for some reason the entry doesn't match expectations.  Perhaps FulcioVerificationException or just RuntimeException?

Reply:
switched to CertificateException